src/flamegraph.cpp

```diff
@@ -274,10 +274,79 @@ QBrush brushImpl(uint hash, BrushType type)
     return brushes.at(hash % brushes.size());
 }
 
-template<typename T>
-QBrush brush(const T& entry, BrushType type)
+QBrush brushBinary(const Data::Symbol& symbol)
+{
+    static QHash<QString, QBrush> brushes;
+
+    auto& brush = brushes[symbol.binary];
+    if (brush == QBrush()) {
+        brush = brushImpl(qHash(symbol.binary), BrushType::Hot);
+    }
+    return brush;
+}
+
+QBrush brushKernel(const Data::Symbol& symbol)
+{
+    static QBrush kernel(QColor(255, 0, 0, 125));
+    static QBrush user(QColor(0, 0, 255, 125));
+
+    if (symbol.isKernel) {
+        return kernel;
+    }
+    return user;
+}
+
+bool isInPathList(const QStringList& paths, const QString& subPath)
 {
-    return brushImpl(qHash(entry), type);
+    auto containsSubPath = [subPath](const auto& path) { return subPath.startsWith(path); };
+
+    return std::any_of(paths.cbegin(), paths.cend(), containsSubPath);
+}
+
+bool isSystemPath(const QString& path)
+{
+    return isInPathList(Settings::instance()->systemPaths(), path);
+}
+
+bool isUserPath(const QString& path)
+{
+    return isInPathList(Settings::instance()->userPaths(), path);
+}
+
+QBrush brushSystem(const Data::Symbol& symbol)
+{
+    static QBrush system(QColor(0, 125, 0, 125));
+    static QBrush user(QColor(200, 200, 0, 125));
+    static QBrush unkown(QColor(50, 50, 50, 125));
+
+    // I have seen [ only on kernel calls
+    if (symbol.path.isEmpty() || symbol.path.startsWith(QLatin1Char('['))) {
+        return unkown;
+    } else if (isSystemPath(symbol.path)) {
+        if (isUserPath(symbol.path)) {
+            return user;
+        }
+        return system;
+    }
+    return user;
+}
+
+template<typename T>
+QBrush brush(const T& entry, Settings::ColorScheme scheme)
```

Comment:
why is this a template? the entry is always a `Symbol`, no?

Reply:
dunno thats not my code, but yes, seems so